File: leetcode18.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        vector<vector<int> > res;
        int len = nums.size();

        for(int i = 0; i < len - 3; i++)
        {
        	for(int j = i+1; j < len - 2; ++j)
        	{
        		int first = j + 1, second = len - 1;
        		while(first < second)
        		{
        			if(nums[i] + nums[j] + nums[first] + nums[second] < target)
        				first++;
        			else if(nums[i] + nums[j] + nums[first] + nums[second] > target)
        				second--;
        			else
        			{
        				vector<int> tmp(4);
        				tmp[0] = nums[i];
        				tmp[1] = nums[j];
        				tmp[2] = nums[first];
        				tmp[3] = nums[second];
        				res.push_back(tmp);

        				first++;
        				while(first < second && nums[first] == nums[first-1])
        					first++;
        				second--;
        				while(second > first && nums[second] == nums[second+1])
        					second--;
        			}
        		}
        		while(j < len - 2 && nums[j] == nums[j+1])
        			j++;
        	}
        	while(i < len - 3 && nums[i] == nums[i+1])
        		i++;
        }
        return res;
    }
};

File: leetcode18.java
public class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        for(int i = 0; i < nums.length-3;i++){
           // List<List<Integer>> tmp = new Array<>();
            res.addAll(threeSum(nums,i+1,target-nums[i]));
            for (int prev = nums[i], next = nums[i+1]; prev == next && i < nums.length-2; prev = next, next = nums[++i + 1]);
        }
        return res;
    }
    
    private List<List<Integer>> threeSum(int[] nums, int start, int target) {
        List<List<Integer>> sums = new ArrayList<>();
        for (int i = start; i < nums.length-2; i++) {
            int first = nums[i];
            if (first * 3 > target) break;
            for (int l = i+1, r = nums.length-1; l < r; ) {
                if (nums[r] * 3 < target) break;
                int sum = first + nums[l] + nums[r];
                if (sum == target) {
                    sums.add(Arrays.asList(nums[start-1], first, nums[l], nums[r]));
                }
                if (sum > target) {
                    while (l < --r && nums[r] == nums[r+1]);
                } else {
                    while (++l < r && nums[l] == nums[l-1]);
                }
            }
            for (int prev = nums[i], next = nums[i+1]; prev == next && i < nums.length-2; prev = next, next = nums[++i + 1]);
        }
        return sums;
    }
}
